src/main/java/com/fasterxml/jackson/core/util/BufferRecyclerPool.java

```diff
@@ -1,17 +1,157 @@
 package com.fasterxml.jackson.core.util;
 
-import com.fasterxml.jackson.core.TokenStreamFactory;
+import java.io.Serializable;
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Interface for entity that controls creation and possible reuse of {@link BufferRecycler}
  * instances used for recycling of underlying input/output buffers.
  *
  * @since 2.16
  */
-public interface BufferRecyclerPool
-    extends java.io.Serializable
-{
-    public BufferRecycler acquireBufferRecycler(TokenStreamFactory forFactory);
+public interface BufferRecyclerPool extends Serializable {
 
-    public void releaseBufferRecycler(BufferRecycler recycler);
+    BufferRecycler acquireBufferRecycler();
+
+    void releaseBufferRecycler(BufferRecycler recycler);
+
+    /**
+     * Returns the default BufferRecyclerPool implementation which is the thread local based one.
+     */
+    static BufferRecyclerPool defaultRecyclerPool() {
+        return PoolStrategy.THREAD_LOCAL.getPool();
+    }
+
+    enum PoolStrategy {
+        NO_OP(new NonRecyclingRecyclerPool()),
+        THREAD_LOCAL(new ThreadLocalRecyclerPool()), 
+        CONCURRENT_DEQUEUE(new ConcurrentDequePool()),
+        LOCK_FREE(new LockFreePool());
+
+        private final BufferRecyclerPool pool;
+
+        PoolStrategy(BufferRecyclerPool pool) {
+            this.pool = pool;
+        }
+
+        public BufferRecyclerPool getPool() {
+            return pool;
+        }
+    }
+
+    /**
+     * Default {@link BufferRecyclerPool} implementation that uses
+     * {@link ThreadLocal} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ThreadLocalRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return BufferRecyclers.getBufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that does not use
+     * any pool but simply creates new instances when necessary.
+     *
+     * @since 2.16
+     */
+    class NonRecyclingRecyclerPool implements BufferRecyclerPool {
+        private static final long serialVersionUID = 1L;
+
+        @Override
+        public BufferRecycler acquireBufferRecycler() {
+            return new BufferRecycler();
+        }
+
+        @Override
+        public void releaseBufferRecycler(BufferRecycler recycler) {
+            ; // nothing to do, relies on ThreadLocal
+        }
+    }
+
+    /**
+     * {@link BufferRecyclerPool} implementation that uses
+     * {@link ConcurrentLinkedDeque} for recycling instances.
+     *
+     * @since 2.16
+     */
+    class ConcurrentDequePool implements BufferRecyclerPool {
```

Comment:
I think it would make sense to require an upper bound for size? Require caller to specify it, even if "unbounded" (Integer.MAX_VALUE).
Would prevent use of `enum` Strategy which I am not big fan of anyway.

Reply:
The [`ConcurrentLinkedDeque`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentLinkedDeque.html) on which I based this implementation is "an unbounded concurrent deque based on linked nodes". 

Moreover it worth to note that the javadoc of that class also says: "Beware that, unlike in most collections, the size method is NOT a constant-time operation. Because of the asynchronous nature of these deques, determining the current number of elements requires a traversal of the elements, and so may report inaccurate results if this collection is modified during traversal."

For these reasons I'm afraid that it won't be possible to give an upper bound size to this specific pool implementation. 

More in general I'm skeptical that such a bounded implementation would be even necessary for this kind of pool. To have its size growing in an unmanageable way it should happen that hundreds of threads try to acquire a `BufferRecycler` at the very same time and then also try to release them all together. In my opinion this would be an extreme and pathological situation and in such unlikely scenario not having a pool at all (using the no-op pool) would be the best possible solution. 

Finally what is not clear to me is how a pool should work when its limit is reached and one more thread tries to release one more `BufferRecycler` into the pool. Should we just throw away that `BufferRecycler` instance? Would this be more efficient or convenient than keeping it into the pool anyway?

Reply:
Ok so the reason I think a limit makes sense is that it reduces unintentional retention of highest number of buffers ever needed concurrently. With thread-bound pooling there's fixed per-thread overhead; but here we can have somewhat poorly bounded number of instances (each thread, or rather read-or-write operation, can request and release any number of recyclers).

So if I am not mistaken, unbounded pool will retain its maximum peak memory usage, ever, as further operation cannot really reduce number pooled (unless they leak BufferRecyclers :) )

And yes, when pool is full, just throwing away `BufferRecycler` would make sense to me.
The goal is to limit amount of memory retained by underlying byte/char buffers, in the worst case scenario.

Finally, if a pool implementation cannot support max size then we wouldn't try to do impossible but rather just document such limitation.

Does this make more sense?

Reply:
In case the limits are in the number of buffers and not the number of bytes I can provide a mpmc array queue from JCTools, which should scale devently under heavy load.

Reply:
Right, since detecting amount of actual memory (via buffers) is difficult to track, I was thinking of simple number of `BufferRecycler`s to retain.

And when doing that, could have access for approximate size. And maybe even `clear()` method if easy enough to do safely (user would need to keep a reference of pool etc; Jackson would not expose any control methods via `BufferRecyclerPool`  (nor size since many implementations cannot provide that efficiently).

Reply:
Can we just keep ConcurrentDequePool and delete the LockFreePool? We only need one of these two pools. Since the new code is extensible, users can provide their own pools anyway.

Reply:
I am ok with that: I think minimally we need:

1. No-recycling "pool"
2. ThreadLocal-based (current impl)
3. New non-thread based pool with maximum size limit

but may also choose to add 

4. New non-thread based pool, unbounded

At the same time I am not super worried about having more than minimal number if implementations are small like they are here.